corporate/lib/stripe.py

```
@@ -5515,38 +5515,71 @@ def check_remote_server_audit_log_data(
-        elif plan.customer.remote_server is not None:
-            remote_server = plan.customer.remote_server
-            billing_session = RemoteServerBillingSession(remote_server=remote_server)
-
-        if plan.fixed_price is not None and not plan.reminder_to_review_plan_email_sent:
-            maybe_send_fixed_price_plan_renewal_reminder_email(plan, billing_session)
-
-        try_to_invoice_plan = True
-        if remote_server:
-            try_to_invoice_plan = check_remote_server_audit_log_data(
-                remote_server, plan, billing_session
-            )
+        if plan.customer.id in failed_customer_ids:  # nocoverage
+            # We've already had a failure for this customer in this
+            # invoicing attempt; skip it so we can process others,
+            # without incorrectly processing other plans on this same
+            # customer.
+            continue
 
-        if try_to_invoice_plan:
-            # plan.next_invoice_date can be None after calling invoice_plan.
-            while plan.next_invoice_date is not None and plan.next_invoice_date <= event_time:
-                billing_session.invoice_plan(plan, plan.next_invoice_date)
-                plan.refresh_from_db()
+        try:
+            review_and_maybe_invoice_plan(plan, event_time)
+        except BillingError as e:
+            failed_customer_ids.add(plan.customer.id)
+            billing_logger.exception(
+                "Invoicing failed: Customer.id: %s, CustomerPlan.id: %s, BillingError: %s",
+                plan.customer.id,
+                plan.id,
+                e.error_description,
+            )
+        except AssertionError:  # nocoverage
+            failed_customer_ids.add(plan.customer.id)
+            billing_logger.exception(
+                "Invoicing failed due to AssertionError: Customer.id: %s, CustomerPlan.id: %s",
+                plan.customer.id,
+                plan.id,
+            )
```

I wasn't sure if we wanted a `except Exception as e:` here as well to catch any other unexpected and otherwise unhandled errors.

---

Probably not essential that we do so, but it would be reasonable.

---

A plausible good pattern is to just catch all exceptions and use `isinstance` checks for logic that wants to vary with the exception class name.

---

That makes sense. Revised for that pattern.